File: src/lib/filingAnchors.ts
export const COPILOT_PURPLE_SHADOW_HOVER = "0 10px 28px -8px rgba(54, 1, 63, 0.38)";

export type FilingAnchor = {
  id: string;
  label: string;
  level: number;
  source: "toc" | "heading" | "target" | "item";
};

const ITEM_RE = /^\s*Item\s+\d{1,2}[A-Z]?\b/i;

/** TOC link text is often only "Item 1." — resolve title from the in-document anchor row. */
function isBareItemTocLabel(label: string): boolean {
  const t = label.replace(/\s+/g, " ").trim();
  if (!/^\s*item\s+\d{1,2}[a-z]?\b/i.test(t)) return false;
  const after = t
    .replace(/^\s*item\s+\d{1,2}[a-z]?\b/i, "")
    .replace(/^[\s.:;\u2014\u00b7\-]+/i, "")
    .trim();
  return after.length < 3;
}

function resolveFragmentTarget(body: HTMLElement, frag: string): Element | null {
  try {
    const byId = body.querySelector(`#${CSS.escape(frag)}`);
    if (byId) return byId;
  } catch {
    /* invalid id selector */
  }
  const byName = body.querySelector(`[name="${escapeAttr(frag)}"]`);
  if (byName) return byName;
  const lower = frag.toLowerCase();
  for (const node of body.querySelectorAll("[id]")) {
    if (node.id.toLowerCase() === lower) return node;
  }
  return null;
}

/** Parse "Item 1. Business …" from a short text blob (table row / cell). */
function parseItemHeadingLine(snippet: string): string | null {
  const s = snippet.replace(/\s+/g, " ").trim();
  const idx = s.search(/\bItem\s+\d{1,2}[A-Za-z]?\b/i);
  if (idx < 0) return null;
  const from = s.slice(idx);
  const m = from.match(/^Item\s+\d{1,2}[A-Za-z]?\b\s*[.:\u2014\u00b7\-]*\s*/i);
  if (!m) return null;
  let rest = from.slice(m[0].length).trim();
  const stop = rest.search(/\bItem\s+\d{1,2}\b/i);
  if (stop >= 0) rest = rest.slice(0, stop).trim();
  rest = rest.replace(/\s*\.{3,}\s*$/, "").trim();
  const prefix = from.slice(0, m[0].length).replace(/\s+/g, " ").trim();
  const head = prefix.replace(/[.:]\s*$/, "").trim();
  if (rest.length >= 2 && /[A-Za-z]{2,}/.test(rest)) {
    return `${head}. ${rest}`.replace(/\s+\./g, ".").slice(0, 220);
  }
  return head.slice(0, 220);
}

function itemHeadingLabelFromFragmentTarget(body: HTMLElement, frag: string): string | null {
  const el = resolveFragmentTarget(body, frag);
  if (!el) return null;
  const row = el.closest("tr");
  const cell = el.closest("td,th");
  const snippet = (row?.textContent ?? cell?.textContent ?? el.textContent ?? "")
    .replace(/\s+/g, " ")
    .trim()
    .slice(0, 1200);
  return parseItemHeadingLine(snippet);
}

function slugifyHeading(input: string): string {
  const base = input
    .trim()
    .toLowerCase()
    .replace(/['"]/g, "")
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
  return base || "section";
}

function decodeFragment(raw: string): string {
  try {
    return decodeURIComponent(raw);
  } catch {
    return raw;
  }
}

function escapeAttr(s: string): string {
  return s.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
}

/** Relative vertical position of `el` within `contentRoot` (0 = top, 1 = bottom). */
export function relativeTopInRoot(el: Element, contentRoot: HTMLElement): number {
  const rr = contentRoot.getBoundingClientRect();
  const er = el.getBoundingClientRect();
  const h = Math.max(contentRoot.scrollHeight, 1);
  return (er.top - rr.top + contentRoot.scrollTop) / h;
}

/**
 * SEC filings duplicate anchors (TOC vs body). Layout is often nested tables; a table with many `#`
 * links can be the real document body — only treat as TOC when the table sits in the upper part
 * of the filing and has several in-doc links.
 */
export function isInsideLikelyToc(el: Element, contentRoot: HTMLElement): boolean {
  const table = el.closest("table");
  if (!table || !contentRoot.contains(table)) return false;
  const hashLinks = table.querySelectorAll('a[href^="#"]');
  if (hashLinks.length < 3) return false;
  const tableTop = relativeTopInRoot(table, contentRoot);
  if (tableTop > 0.42) return false;
  return true;
}

function sortElementsInTreeOrder(nodes: Element[]): Element[] {
  return [...nodes].sort((a, b) => {
    const bit = a.compareDocumentPosition(b);
    if (bit & Node.DOCUMENT_POSITION_FOLLOWING) return -1;
    if (bit & Node.DOCUMENT_POSITION_PRECEDING) return 1;
    return 0;
  });
}

/** Collect all elements matching a fragment (ids often duplicated: TOC vs body). */
function collectAnchorCandidates(contentRoot: HTMLElement, fragment: string): Element[] {
  const raw = fragment.replace(/^#/, "").trim();
  if (!raw) return [];
  const id = decodeFragment(raw);
  const seen = new Set<Element>();
  const add = (node: Element | null) => {
    if (node && contentRoot.contains(node) && !seen.has(node)) {
      seen.add(node);
    }
  };

  try {
    add(contentRoot.querySelector(`#${CSS.escape(id)}`));
  } catch {
    /* ignore */
  }
  add(contentRoot.querySelector(`[name="${escapeAttr(id)}"]`));
  if (id !== raw) {
    try {
      add(contentRoot.querySelector(`#${CSS.escape(raw)}`));
    } catch {
      /* ignore */
    }
    add(contentRoot.querySelector(`[name="${escapeAttr(raw)}"]`));
  }

  const doc = contentRoot.ownerDocument;
  if (doc) {
    const byId = doc.getElementById(id);
    if (byId && contentRoot.contains(byId)) add(byId);
  }

  const lower = id.toLowerCase();
  Array.from(contentRoot.querySelectorAll("[id]")).forEach((node) => {
    if (node.id === id || node.id.toLowerCase() === lower) add(node);
  });

  return sortElementsInTreeOrder(Array.from(seen));
}

/**
 * Prefer the real section over the TOC. SEC HTML almost always repeats the same `id` twice:
 * first hit = TOC row, second hit = in-document anchor. When only one node exists, it is often
 * TOC-only (body uses a different anchor) — return null so the UI can scroll by section label.
 */
export function findBestAnchorTarget(contentRoot: HTMLElement | null, fragment: string): Element | null {
  if (!contentRoot) return null;
  const candidates = collectAnchorCandidates(contentRoot, fragment);
  if (candidates.length === 0) return null;

  if (candidates.length >= 2) {
    return candidates[1];
  }
  return candidates[0];
}

/** @deprecated Use findBestAnchorTarget — kept for quick checks; may return a TOC node. */
export function findAnchorTarget(contentRoot: HTMLElement | null, fragment: string): Element | null {
  return findBestAnchorTarget(contentRoot, fragment);
}

/**
 * Scroll so the target sits near the top of the filing scroll panel.
 * `scrollIntoView` often fails when the only scrollable ancestor is an inner `overflow-y-auto`
 * div (the window does not move, so nothing appears to happen).
 */
export function scrollFilingFragmentIntoView(
  fragment: string,
  contentRoot: HTMLElement | null,
  scrollContainer: HTMLElement | null
) {
  const target = findBestAnchorTarget(contentRoot, fragment);
  if (!target) return;

  if (!scrollContainer || !scrollContainer.contains(target)) {
    target.scrollIntoView({ behavior: "smooth", block: "start" });
    return;
  }

  const margin = 12;
  const cRect = scrollContainer.getBoundingClientRect();
  const tRect = target.getBoundingClientRect();
  const nextTop = scrollContainer.scrollTop + (tRect.top - cRect.top) - margin;
  scrollContainer.scrollTo({ top: Math.max(0, nextTop), behavior: "smooth" });
}

/**
 * After DOMPurify: normalize SEC-style anchors (`name` → `id`), collect jump targets.
 * Order follows the filing: TOC `#` links first (SEC table of contents), then headings,
 * then in-document targets referenced by TOC but missing from the first pass, then Item lines.
 */
export function buildFilingAnchors(doc: Document): FilingAnchor[] {
  const body = doc.body;
  const seen = new Set<string>();
  const anchors: FilingAnchor[] = [];

  const push = (id: string, label: string, level: number, source: FilingAnchor["source"]) => {
    const cleanId = id.trim();
    if (!cleanId || seen.has(cleanId)) return;
    const cleanLabel = label.replace(/\s+/g, " ").trim().slice(0, 220) || cleanId;
    seen.add(cleanId);
    anchors.push({ id: cleanId, label: cleanLabel, level, source });
  };

  body.querySelectorAll("a[name]").forEach((a) => {
    const nm = a.getAttribute("name")?.trim();
    if (nm && !a.id) a.id = nm;
  });

  const linkedFragments = new Set<string>();
  body.querySelectorAll('a[href^="#"]').forEach((a) => {
    const href = a.getAttribute("href")?.trim();
    if (!href || href === "#") return;
    const frag = decodeFragment(href.slice(1));
    if (frag) linkedFragments.add(frag);
  });

  body.querySelectorAll('a[href^="#"]').forEach((a) => {
    const href = a.getAttribute("href")?.trim();
    if (!href || href === "#") return;
    const frag = decodeFragment(href.slice(1));
    if (!frag) return;
    let label = (a.textContent ?? "").replace(/\s+/g, " ").trim() || frag;
    if (isBareItemTocLabel(label)) {
      const rich = itemHeadingLabelFromFragmentTarget(body, frag);
      if (rich) label = rich;
    }
    push(frag, label, 2, "toc");
  });

  const slugSeen = new Map<string, number>();
  body.querySelectorAll("h1,h2,h3,h4,h5,h6").forEach((h) => {
    const label = (h.textContent ?? "").replace(/\s+/g, " ").trim();
    if (!label) return;
    const level = Number(h.tagName.slice(1));
    let id = h.id?.trim() || "";
    if (!id) {
      const base = slugifyHeading(label);
      const prior = slugSeen.get(base) ?? 0;
      slugSeen.set(base, prior + 1);
      id = prior === 0 ? base : `${base}-${prior + 1}`;
      h.id = id;
    }
    push(id, label, level, "heading");
  });

  const labelForUnknownTarget = (el: Element): string => {
    const t = (el.textContent ?? "").replace(/\s+/g, " ").trim();
    if (t.length > 8 && t.length < 200) return t;
    return el.id;
  };

  linkedFragments.forEach((frag) => {
    if (seen.has(frag)) return;
    let el: Element | null = null;
    try {
      el = body.querySelector(`#${CSS.escape(frag)}`);
    } catch {
      el = null;
    }
    if (!el) el = body.querySelector(`[name="${escapeAttr(frag)}"]`);
    if (el) push(frag, labelForUnknownTarget(el), 3, "target");
  });

  const itemSlugSeen = new Map<string, number>();
  body.querySelectorAll("p, td, div, font, span").forEach((el) => {
    const text = (el.textContent ?? "").replace(/\s+/g, " ").trim();
    if (text.length > 350 || text.length < 8 || !ITEM_RE.test(text.slice(0, 120))) return;
    if (el.children.length > 8) return;
    if (el.id && seen.has(el.id)) return;

    const label = text.slice(0, 160);
    const base = slugifyHeading(label.slice(0, 48)) || "item";
    const prior = itemSlugSeen.get(base) ?? 0;
    itemSlugSeen.set(base, prior + 1);
    const newId = prior === 0 ? `sec-item-${base}` : `sec-item-${base}-${prior + 1}`;
    if (!el.id) el.id = newId;
    const finalId = el.id;
    push(finalId, label, 1, "item");
  });

  return anchors.slice(0, 320);
}

File: server/filing_anchors.py
"""Build filing jump targets and extract HTML slices for lazy-loaded sections (SEC 10-K / 10-Q)."""

from __future__ import annotations

import re
from typing import Any, Literal
from urllib.parse import unquote

from bs4 import BeautifulSoup, Tag

FilingAnchorSource = Literal["toc", "heading", "target", "item"]

ITEM_RE = re.compile(r"^\s*Item\s+\d{1,2}[A-Za-z]?\b", re.I)
_BARE_ITEM_TOC = re.compile(r"^\s*item\s+\d{1,2}[a-z]?\b", re.I)
_ITEM_HEAD_LEAD = re.compile(r"^Item\s+\d{1,2}[A-Za-z]?\b\s*[.:\u2014\u00b7\-]*\s*", re.I)
_ITEM_NEXT_BREAK = re.compile(r"\bItem\s+\d{1,2}\b", re.I)


def decode_fragment(raw: str) -> str:
    try:
        return unquote(raw)
    except Exception:
        return raw


def slugify_heading(label: str) -> str:
    s = label.strip().lower().replace("'", "").replace('"', "")
    base = re.sub(r"[^a-z0-9]+", "-", s)
    base = re.sub(r"-+", "-", base).strip("-")
    return base or "section"


def _is_bare_item_toc_label(label: str) -> bool:
    t = re.sub(r"\s+", " ", label).strip()
    if not _BARE_ITEM_TOC.match(t):
        return False
    after = _BARE_ITEM_TOC.sub("", t, count=1)
    after = re.sub(r"^[\s.:;\u2014\u00b7\-]+", "", after, flags=re.I).strip()
    return len(after) < 3


def _resolve_fragment_target(body: Tag, frag: str) -> Tag | None:
    el = body.find(id=frag)
    if el is not None:
        return el
    for node in body.find_all("a", attrs={"name": True}):
        if (node.get("name") or "").strip() == frag:
            return node
    lower = frag.lower()
    for node in body.find_all(id=True):
        if str(node.get("id", "")).lower() == lower:
            return node
    return None


def _parse_item_heading_line(snippet: str) -> str | None:
    s = re.sub(r"\s+", " ", snippet).strip()
    mpos = re.search(r"\bItem\s+\d{1,2}[A-Za-z]?\b", s, re.I)
    if not mpos:
        return None
    from_ = s[mpos.start() :]
    m = _ITEM_HEAD_LEAD.match(from_)
    if not m:
        return None
    rest = from_[m.end() :].strip()
    nb = _ITEM_NEXT_BREAK.search(rest)
    if nb and nb.start() >= 0:
        rest = rest[: nb.start()].strip()
    rest = re.sub(r"\s*\.{3,}\s*$", "", rest).strip()
    prefix = re.sub(r"\s+", " ", m.group(0)).strip()
    head = re.sub(r"[.:]\s*$", "", prefix).strip()
    if len(rest) >= 2 and re.search(r"[A-Za-z]{2,}", rest):
        combined = f"{head}. {rest}"
        combined = re.sub(r"\s+\.", ".", combined)
        return combined[:220]
    return head[:220]


def _item_heading_label_from_fragment_target(body: Tag, frag: str) -> str | None:
    el = _resolve_fragment_target(body, frag)
    if el is None:
        return None
    row = el.find_parent("tr")
    cell = el.find_parent(["td", "th"])
    parts: list[str] = []
    if row is not None and hasattr(row, "get_text"):
        parts.append(row.get_text(" ", strip=True))
    elif cell is not None and hasattr(cell, "get_text"):
        parts.append(cell.get_text(" ", strip=True))
    elif hasattr(el, "get_text"):
        parts.append(el.get_text(" ", strip=True))
    snippet = re.sub(r"\s+", " ", " ".join(parts)).strip()[:1200]
    from_row = _parse_item_heading_line(snippet)
    if from_row and len(from_row) > 12:
        return from_row
    if hasattr(el, "get_text"):
        el_txt = re.sub(r"\s+", " ", el.get_text(" ", strip=True))[:800]
        from_el = _parse_item_heading_line(el_txt)
        return from_row or from_el
    return from_row


_ITEM_KEY_RE = re.compile(r"\bItem\s+(\d{1,2}[A-Za-z]?)\b", re.I)


def _extract_item_key_from_label(label: str) -> str | None:
    m = _ITEM_KEY_RE.search(label)
    if not m:
        return None
    return m.group(1).upper()


def _parse_item_sort_rank(key: str) -> int:
    m = re.match(r"^(\d+)([A-Z]?)$", key, re.I)
    if not m:
        return 999_999
    num = int(m.group(1))
    suf = (m.group(2) or "").upper()
    sub = ord(suf) - 64 if suf else 0
    return num * 100 + sub


def _prefer_richer_item_label(a: str, b: str) -> bool:
    na = re.sub(r"\s+", " ", a).strip()
    nb = re.sub(r"\s+", " ", b).strip()
    if len(na) != len(nb):
        return len(na) > len(nb)
    a_lower = bool(re.search(r"[a-z]", na))
    b_lower = bool(re.search(r"[a-z]", nb))
    if a_lower and not b_lower:
        return True
    if not a_lower and b_lower:
        return False
    return na.casefold() < nb.casefold()


def _merge_item_anchor_group(group: list[dict[str, Any]]) -> dict[str, Any]:
    best = group[0]
    for g in group[1:]:
        if _prefer_richer_item_label(g["label"], best["label"]):
            best = g
    return best


def _dedupe_and_sort_item_anchors(anchors: list[dict[str, Any]]) -> list[dict[str, Any]]:
    non_items: list[dict[str, Any]] = []
    buckets: dict[str, list[dict[str, Any]]] = {}
    for a in anchors:
        key = _extract_item_key_from_label(a["label"])
        if key:
            buckets.setdefault(key, []).append(a)
        else:
            non_items.append(a)
    merged_items: list[dict[str, Any]] = []
    for group in buckets.values():
        merged_items.append(_merge_item_anchor_group(group))

    def _sort_key(x: dict[str, Any]) -> int:
        k = _extract_item_key_from_label(x["label"])
        return _parse_item_sort_rank(k) if k else 999_999

    merged_items.sort(key=_sort_key)
    return non_items + merged_items


def build_filing_anchors(raw_html: str, max_anchors: int = 320) -> list[dict[str, Any]]:
    soup = BeautifulSoup(raw_html, "html.parser")
    body = soup.body
    if body is None:
        return []

    seen: set[str] = set()
    anchors: list[dict[str, Any]] = []

    def push(anchor_id: str, label: str, level: int, source: FilingAnchorSource) -> None:
        clean_id = anchor_id.strip()
        if not clean_id or clean_id in seen:
            return
        clean_label = re.sub(r"\s+", " ", label).strip()[:220] or clean_id
        seen.add(clean_id)
        anchors.append({"id": clean_id, "label": clean_label, "level": level, "source": source})

    for a in body.select("a[name]"):
        nm = (a.get("name") or "").strip()
        if nm and not a.get("id"):
            a["id"] = nm

    linked_fragments: set[str] = set()
    for a in body.select('a[href^="#"]'):
        href = (a.get("href") or "").strip()
        if not href or href == "#":
            continue
        frag = decode_fragment(href[1:])
        if frag:
            linked_fragments.add(frag)

    for a in body.select('a[href^="#"]'):
        href = (a.get("href") or "").strip()
        if not href or href == "#":
            continue
        frag = decode_fragment(href[1:])
        if not frag:
            continue
        label = re.sub(r"\s+", " ", (a.get_text() or "")).strip() or frag
        if _is_bare_item_toc_label(label):
            rich = _item_heading_label_from_fragment_target(body, frag)
            if rich:
                label = rich
        push(frag, label, 2, "toc")

    slug_seen: dict[str, int] = {}
    for h in body.select("h1,h2,h3,h4,h5,h6"):
        label = re.sub(r"\s+", " ", (h.get_text() or "")).strip()
        if not label:
            continue
        try:
            level = int(h.name[1])
        except (TypeError, ValueError):
            level = 2
        hid = (h.get("id") or "").strip()
        if not hid:
            base = slugify_heading(label[:200])
            prior = slug_seen.get(base, 0)
            slug_seen[base] = prior + 1
            hid = base if prior == 0 else f"{base}-{prior + 1}"
            h["id"] = hid
        push(hid, label, level, "heading")

    def label_for_unknown_target(el: Tag) -> str:
        t = re.sub(r"\s+", " ", (el.get_text() or "")).strip()
        if 8 < len(t) < 200:
            return t
        return str(el.get("id") or "")

    for frag in linked_fragments:
        if frag in seen:
            continue
        el = body.find(id=frag)
        if el is None:
            for node in body.find_all("a", attrs={"name": True}):
                if (node.get("name") or "").strip() == frag:
                    el = node
                    break
        if el is None:
            lower = frag.lower()
            for node in body.find_all(id=True):
                if str(node.get("id", "")).lower() == lower:
                    el = node
                    break
        if el is not None:
            push(frag, label_for_unknown_target(el), 3, "target")

    item_slug_seen: dict[str, int] = {}
    for el in body.select("p, td, div, font, span"):
        if not isinstance(el, Tag):
            continue
        text = re.sub(r"\s+", " ", (el.get_text() or "")).strip()
        if len(text) > 350 or len(text) < 8:
            continue
        if not ITEM_RE.match(text[:120]):
            continue
        if len(list(el.children)) > 8:
            continue
        eid = (el.get("id") or "").strip()
        if eid and eid in seen:
            continue
        label = text[:160]
        base = slugify_heading(label[:48]) or "item"
        prior = item_slug_seen.get(base, 0)
        item_slug_seen[base] = prior + 1
        new_id = f"sec-item-{base}" if prior == 0 else f"sec-item-{base}-{prior + 1}"
        if not el.get("id"):
            el["id"] = new_id
        final_id = str(el.get("id") or new_id)
        push(final_id, label, 1, "item")

    merged = _dedupe_and_sort_item_anchors(anchors)
    return merged[:max_anchors]


def extract_fragment_html(raw_html: str, fragment: str, max_chars: int) -> str | None:
    """
    Return a large HTML slice starting at the first tag that defines this fragment (id or name).
    SEC filings nest anchors in tables; a raw substring is more reliable than sibling walking.
    """
    fid = decode_fragment(fragment.strip().lstrip("#"))
    if not fid:
        return None

    esc = re.escape(fid)
    patterns = [
        rf'\b(?:id|name)\s*=\s*"{esc}"',
        rf"(?:id|name)\s*=\s*'{esc}'",
    ]
    pos: int | None = None
    for pat in patterns:
        m = re.search(pat, raw_html, re.I)
        if m:
            pos = m.start()
            break
    if pos is None:
        for m in re.finditer(r'(?i)\b(?:id|name)\s*=\s*["\']([^"\']+)["\']', raw_html):
            if m.group(1).strip().lower() == fid.lower():
                pos = m.start()
                break
    if pos is None:
        return None

    start = raw_html.rfind("<", 0, pos)
    if start < 0:
        start = pos
    return raw_html[start : start + max_chars]
